lib/imagefs.coffee

```diff
@@ -23,16 +23,71 @@ _ = require('lodash')
 filedisk = require('file-disk')
 fs = Promise.promisifyAll(require('fs'))
 replaceStream = require('replacestream')
+ext2fs = require('ext2fs')
+
 driver = require('./driver')
+utils = require('./utils')
 
 checkImageType = (image) ->
 	if not (_.isString(image) or image instanceof filedisk.Disk)
 		throw new Error('image must be a String (file path) or a filedisk.Disk instance')
 
+composeDisposers = (outerDisposer, createInnerDisposer) ->
+	Promise.resolve(outerDisposer)
+	.then (outerDisposer) ->
+		outerDisposer._promise
+		.then (outerResult) ->
+			Promise.resolve(createInnerDisposer(outerResult))
+			.then (innerDisposer) ->
+				innerDisposer._promise
+				.then (innerResult) ->
+					Promise.resolve(innerResult)
+					.disposer (innerResult) ->
+						Promise.resolve(innerDisposer._data(innerResult))
+						.then ->
+							outerDisposer._data(outerResult)
+			.catch (err) ->
+				outerDisposer._data(outerResult)
+				throw err
```

Comment:
* There is no `innerDisposer._data(outerResult)`, it is `innerDisposer._data(innerResult)`: we always dispose what we created.
 * the catch is there in case the innerDisposer fails to create its resource for some reason, so the outerDisposer's resources are disposed anyway.

Imagine you have 2 disposers:

```coffeescript
mountFilesystem = (diskImagePath) ->
    functionThatTakesAFileAndReturnsAPromiseOfAFilesystem(diskImagePath)
    .disposer (fs) ->
        fs.umountAsync()
```

and


```coffeescript
useFile = (fs, path, mode) ->
    fs.openAsync(path, mode)
    .disposer (fd) ->
        fs.closeAsync(fd)
```

The normal way of reading a file would be

```coffeescript
Promise.using mountFilesystem('my_disk_image'), (fs) ->
    Promise.using useFile(fs, 'some_file_in_the_fs', 'r'), (fd) ->
        fs.readFileAsync(fd, 'utf8')
        .then(console.log)
```

But now imagine you want to create a function that returns a disposer of a file on the disk image. Something that could be used like this:

```coffeescript
Promise.using useFileOnDisk('my_disk_image', 'my_file_in_the_image', 'r'), (fd) ->
    fs.readFileAsync(fd, 'utf8')
    .then(console.log)
```

Let's try to implement it:

```coffeescript
useFileOnDisk = (diskImagePath, filePathInTheImage, mode) ->
    Promise.using mountFilesystem('my_disk_image'), (fs) ->
        Promise.using useFile(fs, 'some_file_in_the_fs', 'r'), (fd) ->
            fd
```

This doesn't work because when the caller gets the fd, the file is closed and the disk image too.
So we need a way of combining both disposers. That's what `composeDisposers` does.

I agree the code is difficult to read. The idea is:
1) allocate outerDisposer's resources
2) allocate innerDisposer's resources (using outerDisposer's result)
3) use innerDisposer's resources \o/
4) dispose innerDisposer's resources
5) dispose outerDisposer's resources

The `catch` is basically "If step 2. fails, goto 5"